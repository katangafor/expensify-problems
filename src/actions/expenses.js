import database from '../firebase/firebase';

// ADD_EXPENSE
// this one parameter is an object, which has default parameters if fed an object, and defaults to an empty object if fed no object
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
})

export const startAddExpense = (expenseData = {}) => {
  // this part only works cause of redux-thunk middleware
  return (dispatch) => {
    const {
      description = '', 
      note = '', 
      amount = 0, 
      createdAt = 0 
    } = expenseData;
    const expense = { description, note, amount, createdAt };

    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }))
    })
  }
}

// REMOVE EXPENSE
export const removeExpense = ({ id }) => ({
  type: 'REMOVE_EXPENSE',
  id
})

// START REMOVE EXPENSES
export const startRemoveExpense = ({ id }) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({ id }))
    })
  }
}

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})

// SET EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

// START SET EXPENSES
export const startSetExpenses = (expenses) => {
  return (dispatch) => {
    return database.ref('expenses').once('value').then((snapshot) => {
      const expenses = [];
      snapshot.forEach((childSnapshot) => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        })
      })
      dispatch(setExpenses(expenses))
    })
  }
}