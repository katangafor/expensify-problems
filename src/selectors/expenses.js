

// Takes in the filters and the sort criteria, and reeturns a new array that's properly sorted
export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses.filter((expense) => {
    // this is probably better done with momment() methods
    const startDateMatch = startDate !== undefined ? expense.createdAt >= startDate : true;
    const endDateMatch = endDate !== undefined ? expense.createdAt <= endDate : true;
    const textMatch = text ? expense.description.toLowerCase().includes(text.toLowerCase()) : true;

    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1;
    }
  });
}