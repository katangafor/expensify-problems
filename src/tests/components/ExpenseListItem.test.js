import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';

// doesn't the ExpenseListItem take a prop? Is that not what this spread expenses object is? Why doesn't it work when I give it a name?
// Cause it doesn't take single object prop, but four props?
test('should render an expense list item', () => {
  const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>);
  expect(wrapper).toMatchSnapshot();
})