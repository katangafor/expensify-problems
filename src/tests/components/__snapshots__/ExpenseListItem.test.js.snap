// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render an expense list item 1`] = `
<div>
  <Link
    to="/edit/1a"
  >
    <h3>
      rocket league
    </h3>
  </Link>
  <p>
    $1.00
    -
    Dec, 31st, 1969
  </p>
</div>
`;
